### PR TITLE
Remove struct dep

### PR DESCRIPTION
**Pull request by lipp**
Credits go to @moteus THX!

### CODE REVIEW

File: src/websocket/frame.lua [hunk 1]
```diff
@@ -1,158 +1,202 @@
...
+      pos, payload = read_int16(encoded, pos)
+    elseif payload == 127 then
+      if left < 8 then
+        return nil, 8 - left, nil, start
+      end
+      local high, low
+      pos, high = read_int32(encoded, pos)
+      pos, low  = read_int32(encoded, pos)
+      payload = high * 2^32 + low
+      if payload < 0xFFFF or payload > 2^53 then
```
Comment (lipp) [inline src/websocket/frame.lua]: @moteus  should'nt this be payload < 0xFF ?

File: src/websocket/frame.lua [hunk 2]
```diff
@@ -1,158 +1,202 @@
...
+        assert(false, 'INVALID PAYLOAD '..payload)
```
Comment (lipp) [inline src/websocket/frame.lua]: don't assert, in "critical" situations call `error('invalid payload)` to exit function

File: src/websocket/frame.lua [hunk 3]
```diff
@@ -1,158 +1,202 @@
...
     end
-    local xored = schar(unpack(transformed,1,#original))
-    tinsert(transformed_arr,xored)
+
+    decoded = string.sub(encoded, pos, pos + payload - 1)
+    pos = pos + payload
+  end
+
+  local fin    = bit.band(header, bit_7) ~= 0
+  local rsv1   = bit.band(header, bit_6) ~= 0
```
Comment (lipp) [inline src/websocket/frame.lua]: check all bits at once: 

```lua
local rsv = bit.band(header, bit_4_5_6)
if rsv ~= 0 then
  error('reserved bits must not be set')
end
```

File: src/websocket/frame.lua [hunk 4]
```diff
@@ -1,158 +1,202 @@
...
+  local rsv2   = bit.band(header, bit_5) ~= 0
+  local rsv3   = bit.band(header, bit_4) ~= 0
+  local opcode = bit.band(header, bit_0_3)
+
+  return decoded,fin,opcode,pos,masked,rsv1,rsv2,rsv3
```
Comment (lipp) [inline src/websocket/frame.lua]: don't return rsv values. we don't need them (error is thrown if they are set)

File: src/websocket/frame.lua [hunk 5]
```diff
@@ -1,158 +1,202 @@
...
+end
+
+local decode = function(encoded)
+  local decoded, fin, opcode, pos, masked, rsv1, rsv2, rsv3 = decode_by_pos(encoded, 1)
+  local rest
+  if decoded then
+    local rest = encoded:sub(pos)
+    return decoded, fin, opcode, rest, masked, rsv1, rsv2, rsv3
```
Comment (lipp) [inline src/websocket/frame.lua]: don't return rsv

**Comment by lipp** [comment]
@moteus all calls to frame.decode must be wrapped in `pcall` and we need to close connection
when decode throws.

File: src/websocket/frame.lua (hunk 1 above)
Comment (moteus) [inline src/websocket/frame.lua]: No. In fact this is overflow check. I really do not think that some one whant send frame over 2G.

File: src/websocket/frame.lua (hunk 2 above)
Comment (moteus) [inline src/websocket/frame.lua]: I prefer do not use error eather but just return error code/message from this function.

File: src/websocket/frame.lua (hunk 3 above)
Comment (moteus) [inline src/websocket/frame.lua]: Raise error is not good idea. This bits uses to implement extesion. So better just returns them and allow hier level to handle this.

File: src/websocket/frame.lua (hunk 2 above)
Comment (lipp) [inline src/websocket/frame.lua]: this function already has plenty of return values. what should the caller do (in case of >2gb) anyhow, but close the connection?

File: src/websocket/frame.lua (hunk 3 above)
Comment (lipp) [inline src/websocket/frame.lua]: as long as lua-websockets has no support for extensions i'd stick to the rfc:

  RSV1, RSV2, RSV3:  1 bit each

      MUST be 0 unless an extension is negotiated that defines meanings
      for non-zero values.  If a nonzero value is received and none of
      the negotiated extensions defines the meaning of such a nonzero
      value, the receiving endpoint MUST _Fail the WebSocket
      Connection_.

Comment (moteus) [inline src/websocket/frame.lua]: Yes. But there no reason to crash entire server to close connection. Decoder just returns rsv. and server terminate connection with protocol error. As I said I think decoder should never raise error. Decoder can not know eather server supports extension or not and that decision should make server code.

File: src/websocket/frame.lua (hunk 2 above)
Comment (moteus) [inline src/websocket/frame.lua]: I do not change this bihavior. The problem is now frame decoder can not return error and to allow this there need update backend code as whell. I think it should be done in another PR

File: src/websocket/frame.lua (hunk 3 above)
Comment (lipp) [inline src/websocket/frame.lua]: i dont want to crash the server! :)
the websocket connection should be closed, regardless if the decode fails in the client or the server.
i dont care too much if this is implement via throwing an error or passing values around.
the decode function already has plenty of return values and i dont like the number to grow as this
gets unreadable. 
but again: the main thing is (as long as lua-websockets does not support extensions) to close the connection. you brought up the point about conformance ;)

**Comment by lipp** [comment]
replaced with #79
